src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs

```diff
@@ -694,6 +700,46 @@ protected override void ComputeDependencyNodeDependencies(List<DependencyNodeCor
                                 CorInfoImpl.IsMethodCompilable(this, methodCodeNodeNeedingCode.Method))
                                 _methodsWhichNeedMutableILBodies.Add(ecmaMethod);
                         }
+                        if (method.IsAsyncThunk())
+                        {
+                            // The synthetic async thunks require references to methods/types that may not have existing methodRef entries in the version bubble.
+                            // These references need to be added to the mutable module if they don't exist.
+                            var requiredMutableModuleReferences = method.IsAsyncVariant() ?
+                                AsyncThunkILEmitter.GetRequiredReferencesForAsyncThunk(((AsyncMethodVariant)method).Target)
+                                : AsyncThunkILEmitter.GetRequiredReferencesForTaskReturningThunk(method);
```

Comment:
What is the set of methods/types this returns? Would we be able to obtain the same set of types/method by inspecting the IL of `method`?

What I'm getting at is whether we'd be able to leave AsyncThunks.cs as-is and instead here do `ilProvider.GetMethodIL(method)`, construct an `ILReader` over the body, do a for loop over all instructions and extract the dependencies from there. I'm concerned about the readability of AsyncThunks.cs basically.

Reply:
The approach in your PR is definitely faster than parsing IL, but parsing IL stream is correct by construction. I don't think we need to squeeze out all the throughput from this; I'd go with the correct-by-construction approach if it's an option.

The loop would look similar to this:

https://github.com/dotnet/runtime/blob/55728d4db73366d1f1b276f8e60eb5581163751b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/FlowAnnotations.cs#L684-L708

except we'd handle call/callvirt/etc (and probably the references from the exception handling info of the MethodIL).